### PR TITLE
Bug 2082667: Separate controller for the node draining

### PR DESCRIPTION
**Pull request by lobziik**
Introduces new controller which responsible for node draining.
Basically drain logic was moved out to separate controller, running by the same manager. Now machine-controller will block machine deletion until "Drained" condition will not be set to "True" by the "machine-drain-controller". Logic there was mostly untouched and extracted as is, except some events emission was added.

Tested manually on vSphere.

### CODE REVIEW

File: pkg/controller/machine/controller.go [hunk 1]
```diff
@@ -217,30 +221,12 @@ func (r *ReconcileMachine) Reconcile(ctx context.Context, request reconcile.Requ
...
-					machinev1.ConditionSeverityWarning,
-					"could not drain machine: %v", err,
-				))
-				return delayIfRequeueAfterError(err)
-			}
-			conditions.Set(m, conditions.TrueCondition(machinev1.MachineDrained))
+		// check if machine was already drained
+		drainedCondition := conditions.Get(m, machinev1.MachineDrained)
+		if drainedCondition == nil || drainedCondition.Status != corev1.ConditionTrue {
+			klog.Infof("%v: wait for node draining to proceed deletion", machineName)
```
Comment (JoelSpeed) [inline pkg/controller/machine/controller.go:227]: Nit: Minor wording suggestion
```suggestion
			klog.Infof("%s: waiting for node to be drained before deleting instance", machineName)
```

File: pkg/controller/machine/drain_controller.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+		} else {
+			d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainSkipped", "Node drain skipped")
+			drainFinishedCondition.Message = "Node drain skipped"
+		}
+
+		d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainFinished", "Node drain finished")
+		conditions.Set(m, drainFinishedCondition)
+		// requeue request in case of failed update
+		err := d.Client.Status().Update(ctx, m)
+		return reconcile.Result{}, err
```
Comment (JoelSpeed) [inline pkg/controller/machine/drain_controller.go:95]: Not a fan of naked error returns like this, can we wrap it with context, adding unique context makes it easier to find where it came from when debugging
```suggestion
		if err := d.Client.Status().Update(ctx, m); err != nil {
		  return reconcile.Result{}, fmt.Errorf("could not update machine status: %w", err)
		}
		
		return reconcile.Result{}, nil
```

**Review by JoelSpeed — COMMENTED** [review]
Can we look into adding some dedicated testing for the drain controller as presently we don't seem to have any, some basic input/output of the reconcile using envtest would be good to check it doesn't modify machines that aren't deleted, and sets appropriate conditions otherwise

File: pkg/controller/machine/drain_controller.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+				conditions.Set(m, conditions.FalseCondition(
+					machinev1.MachineDrained,
+					machinev1.MachineDrainError,
+					machinev1.ConditionSeverityWarning,
+					"could not drain machine: %v", err,
+				))
+				d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainRequeued", "Node drain requeued: %v", err.Error())
+				return delayIfRequeueAfterError(err)
+			}
+			d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainSucceeded", "Node drain succeeded")
```
Comment (JoelSpeed) [inline pkg/controller/machine/drain_controller.go:84]: There is an event at the end of `drainNode` which means we get two events every time this completes successfully, do we want to remove one of them?

File: pkg/controller/machine/drain_controller.go [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+	if !m.ObjectMeta.DeletionTimestamp.IsZero() && *m.Status.Phase == phaseDeleting && !alreadyDrained {
+		drainFinishedCondition := conditions.TrueCondition(machinev1.MachineDrained)
+
+		if _, exists := m.ObjectMeta.Annotations[ExcludeNodeDrainingAnnotation]; !exists && m.Status.NodeRef != nil {
+			// pre-drain.delete lifecycle hook
+			// Return early without error, will requeue if/when the hook owner removes the annotation.
+			if len(m.Spec.LifecycleHooks.PreDrain) > 0 {
+				klog.Infof("%v: not draining machine: lifecycle blocked by pre-drain hook", m.Name)
+				d.eventRecorder.Eventf(m, corev1.EventTypeNormal, "DrainBlocked", "Drain blocked by pre-drain hook")
```
[the remark this answers is not in the document]
Comment (lobziik) [reply pkg/controller/machine/drain_controller.go:69]: I don't see a necessity, to be honest. We treat lack of condition as false, and do not proceed with machine deletion in this case.

**Comment by JoelSpeed** [comment]
/lgtm cancel

The machine controller is panicking, something isn't initialised properly, https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_machine-api-operator/1023/pull-ci-openshift-machine-api-operator-master-e2e-vsphere-operator/1532667754660761600/artifacts/e2e-vsphere-operator/gather-extra/artifacts/pods/openshift-machine-api_machine-api-controllers-d5d46c8d8-4nhp8_machine-controller_previous.log

```
goroutine 518 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/src/github.com/openshift/machine-api-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:118 +0x1f4
panic({0x1bcc880, 0x3364a20})
	/usr/lib/golang/src/runtime/panic.go:838 +0x207
github.com/openshift/machine-api-operator/pkg/controller/machine.(*machineDrainController).Reconcile(0xc0003d4090, {0x22a6e40, 0xc000f40720}, {{{0xc00002c978?, 0x10?}, {0xc0007cf5f0?, 0x40d847?}}})
	/go/src/github.com/openshift/machine-api-operator/pkg/controller/machine/drain_controller.go:61 +0x1d8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x22a6d98?, {0x22a6e40?, 0xc000f40720?}, {{{0xc00002c978?, 0x1e94060?}, {0xc0007cf5f0?, 0x4041f4?}}})
	/go/src/github.com/openshift/machine-api-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:121 +0xc8
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000322640, {0x22a6d98, 0xc00092f780}, {0x1cdf540?, 0xc000379a20?})
	/go/src/github.com/openshift/machine-api-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:320 +0x33c
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000322640, {0x22a6d98, 0xc00092f780})
	/go/src/github.com/openshift/machine-api-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:273 +0x1d9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/src/github.com/openshift/machine-api-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:234 +0x85
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
	/go/src/github.com/openshift/machine-api-operator/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:230 +0x325
```

Looks like line 61 of drain controller needs to protect against a nil phase pointer

**Comment by lobziik** [comment]
/hold cancel

Checked vsphere machine controller logs in e2e-vsphere-operator, drainer controller works well, as far as i can tell.

**Comment by JoelSpeed** [comment]
/override ci/prow/e2e-vsphere-serial

The test failure is related to CSI, everything to do with this PR seems to be working ok, no need to wait on this

**Pull request merged**